src/satosa/micro_services/custom_routing.py

```diff
@@ -2,14 +2,117 @@
 from base64 import urlsafe_b64encode
 
 from satosa.context import Context
+from satosa.internal import InternalData
+
 from .base import RequestMicroService
 from ..exception import SATOSAConfigurationError
 from ..exception import SATOSAError
+from ..exception import SATOSAStateError
 
 
 logger = logging.getLogger(__name__)
 
 
+class CustomRoutingError(SATOSAError):
+    """
+    SATOSA exception raised by CustomRouting rules
+    """
+    pass
+
+
+class DecideBackendByTarget(RequestMicroService):
+    """
+    Select which backend should be used based on who is the SAML IDP
+    """
+
+    def __init__(self, config, *args, **kwargs):
+        """
+        Constructor.
+        :param config: microservice configuration loaded from yaml file
+        :type config: Dict[str, Dict[str, str]]
+        """
+        super().__init__(*args, **kwargs)
+        self.target_mapping = config['target_mapping']
+        self.endpoint_paths = config['endpoint_paths']
+        self.default_backend = config['default_backend']
+
+        if not isinstance(self.endpoint_paths, list):
+            raise SATOSAConfigurationError()
+
+    def register_endpoints(self):
+        """
+        URL mapping of additional endpoints this micro service needs to register for callbacks.
+
+        Example of a mapping from the url path '/callback' to the callback() method of a micro service:
+            reg_endp = [
+                ("^/callback1$", self.callback),
+            ]
+
+        :rtype List[Tuple[str, Callable[[satosa.context.Context, Any], satosa.response.Response]]]
+
+        :return: A list with functions and args bound to a specific endpoint url,
+                 [(regexp, Callable[[satosa.context.Context], satosa.response.Response]), ...]
+        """
+
+        # this intercepts disco response
+        return [
+            (path , self.backend_by_entityid)
+            for path in self.endpoint_paths
+        ]
+
+    def _get_backend(self, context:str, entity_id:str) -> str:
```

Comment:
`context` is not a `str`, but a `satosa.context.Context` object.
It is also not used here, so we could even skip it.

```suggestion
    def _get_backend(self, context:Context, entity_id:str) -> str:
```